coordinator/src/contract.rs

```diff
@@ -0,0 +1,215 @@
+use dlc_manager::ContractId;
+use dlc_manager::ReferenceId;
+use ln_dlc_node::node::rust_dlc_manager::DlcChannelId;
+use rust_decimal::prelude::FromPrimitive;
+use rust_decimal::Decimal;
+use time::OffsetDateTime;
+use trade::cfd::calculate_margin;
+use trade::cfd::calculate_pnl;
+use trade::Direction;
+use uuid::Uuid;
+
+pub struct ContractTransaction {
+    pub id: Uuid,
+    pub timestamp: OffsetDateTime,
+    pub channel_id: DlcChannelId,
+    pub contract_id: ContractId,
+    pub trader: PublicKey,
+    pub transaction_state: ContractTransactionState,
+}
+
+pub struct TradeParams {
+    pub contract_txid: Uuid,
+    pub trader: PublicKey,
+    pub quantity: f32,
+    pub leverage: f32,
+    pub average_price: f32,
+    pub direction: Direction,
+}
+
+pub enum ContractTransactionState {
+    Pending,
+    Success,
+    Failed,
+}
+
+pub struct ContractExecutor {
+    pool: Pool<ConnectionManager<PgConnection>>,
+}
+
+impl ContractExecutor {
+    pub fn new(pool: Pool<ConnectionManager<PgConnection>>) -> Self {
+        ContractExecutor { pool }
+    }
+
+    /// Starts a contract transaction, by creating a new contract transaction and temporarily stores
+    /// the trade params.
+    ///
+    /// Returns an uniquely generated contract txid as [`dlc_manager::ReferenceId`]
+    pub fn start_contract_transaction(
+        &self,
+        contract_txid: ReferenceId,
+        previous_txid: Option<ReferenceId>,
+        contract_id: ContractId,
+        channel_id: DlcChannelId,
+        trade_params: &commons::TradeParams,
+    ) -> Result<()> {
+        let mut conn = self.pool.get()?;
+        conn.transaction(|conn| {
+            db::contract_transactions::create(
+                conn,
+                contract_txid,
+                previous_txid,
+                contract_id,
+                channel_id,
+                &trade_params.pubkey,
+            )?;
+            db::trade_params::insert(conn, contract_txid, trade_params)?;
+
+            diesel::result::QueryResult::Ok(())
+        })?;
+
+        Ok(())
+    }
+
+    pub fn fail_contract_transaction(&self, contract_txid: ReferenceId) -> Result<()> {
+        let mut conn = self.pool.get()?;
+        db::contract_transactions::set_contract_transaction_state_to_failed(
+            &mut conn,
+            contract_txid,
+        )?;
+
+        Ok(())
+    }
+
+    /// Completes the contract transaction as successful and updates the 10101 meta data
+    /// accordingly in a single database transaction.
+    /// - Set contract transaction to success
+    /// - If not closing: Updates the `[PostionState::Proposed`] position state to
+    ///   `[PostionState::Open]`
+    /// - If closing: Calculates the pnl and sets the `[PostionState::Closing`] position state to
+    ///   `[PostionState::Closed`]
+    /// - Creates and inserts the new trade
+    pub fn finish_contract_transaction(
+        &self,
+        contract_txid: ReferenceId,
+        closing: bool,
```

Comment:
I opted not to do that, because its a refactoring on the position state, that we want to get rid off. Hence I wanted to make it a bit ugly here, because this should get removed once we get rid of the position state :)